### PR TITLE
Experimental support for video player

### PR DESCRIPTION
**Pull request by ggichure**
## Pull Request Description

 -  [youtube_player_flutter](https://pub.dev/packages/youtube_player_flutter) to play  yt videos on android and ios
 -  [youtube_player_iframe](https://pub.dev/packages/youtube_player_iframe) to play  videos on web , etc
 -  [river_player](https://pub.dev/packages/river_player) to play any file with a standard codec `mp4,mov` etc 
 

## Issue Being Fixed

Issue Number: #383

## Screenshots / Recordings

https://github.com/thunder-app/thunder/assets/31619962/8e5ecf34-72bb-44ee-b31d-626bbe36310e

## Checklist

- [x] Did you update CHANGELOG.md?
- [x] Did you use localized strings where applicable?
- [x] Did you add `semanticLabel`s where applicable for accessibility?

### CODE REVIEW

**Comment by hjiangsu** [comment]
I know this is in draft at the moment, but I'd like to just chime in that there is the option to only add partial support for videos in this PR!

We can extend video player functionality in future PRs to support edge cases 😊

File: lib/core/enums/video_auto_play.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+enum VideoAutoPlay {
+  never(label: 'Never'),
```
Comment (micahmo) [inline lib/core/enums/video_auto_play.dart:2]: Can you see about loading these strings from localization? You might have to add a property/method which does a dynamic lookup of the resources using the global context.

File: lib/core/enums/video_playback_speed.dart [hunk 2]
```diff
@@ -0,0 +1,16 @@
+enum VideoPlayBackSpeed {
+  pointTow5x(label: '0.25'),
+  point5x(label: '0.5'),
+  pointSeven5x(label: '0.75'),
+  normal(label: '1'),
```
Comment (micahmo) [inline lib/core/enums/video_playback_speed.dart:5]: What do you think about labeling this with `x`, like `1x`, `2x`, etc.?

File: lib/core/enums/local_settings.dart [hunk 3]
```diff
@@ -11,7 +11,8 @@ enum LocalSettingsCategories {
   account('Account'),
   about('About'),
   debug('Debug'),
-  theming('Theming');
+  theming('Theming'),
+  videoPlayer('videoPlayer');
```
Comment (micahmo) [inline lib/core/enums/local_settings.dart:15]: It looks like this string is used directly in search results. It should probably be translated, but for now, can you stylize it in a more readable way, like `Video Player`?

File: lib/l10n/app_en.arb [hunk 4]
```diff
@@ -2263,6 +2263,30 @@
...
+  },
+  "videoAutoMute": "Mute Videos",
+  "@videoAutoMute": {
+    "description": "Option to always mute videos"
+  },
+  "videoAutoPlay": "Video Autoplay",
+  "@videoAutoPlay": {
+    "description": "Option to always autoplay videos"
+  },
+  "videoDefaultPlaybackSpeed": "Default Playback speed",
```
Comment (micahmo) [inline lib/l10n/app_en.arb:2286]: Can you capitalize Speed?

File: lib/settings/pages/video_player_settings.dart [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+class _VideoPlayerSettingsPageState extends State<VideoPlayerSettingsPage> {
+  /// Loading
+  bool isLoading = true;
+
+  LocalSettings? settingToHighlight;
+  GlobalKey settingToHighlightKey = GlobalKey();
+
+  /// Toggle to always start video in fullscreen landscape when enabled
+  bool videoAutoFullscreen = true;
```
Comment (micahmo) [inline lib/settings/pages/video_player_settings.dart:30]: Some of these default don't seem the match the way you've set them elsewhere (i.e., ThunderBloc).

File: lib/settings/pages/video_player_settings.dart [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                    iconEnabled: Icons.loop,
+                    iconDisabled: Icons.loop_outlined,
+                    onToggle: (bool value) => setPreferences(LocalSettings.videoAutoLoop, value),
+                    highlightKey: settingToHighlight == LocalSettings.videoAutoLoop ? settingToHighlightKey : null,
+                  ),
+                  ListOption(
+                    description: l10n.videoAutoPlay,
+                    value: ListPickerItem(label: videoAutoPlay.name, icon: Icons.video_settings_outlined, payload: videoAutoPlay),
+                    options: [
+                      ListPickerItem(label: VideoAutoPlay.never.label, payload: VideoAutoPlay.never),
```
Comment (micahmo) [inline lib/settings/pages/video_player_settings.dart:158]: Can you add icons for each of these items?

File: lib/settings/pages/video_player_settings.dart [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                    ],
+                    icon: Icons.play_circle,
+                    onChanged: (value) async => setPreferences(LocalSettings.videoAutoPlay, value.payload.name),
+                    highlightKey: settingToHighlight == LocalSettings.videoAutoPlay ? settingToHighlightKey : null,
+                  ),
+                  ListOption(
+                    description: l10n.videoDefaultPlaybackSpeed,
+                    value: ListPickerItem(label: videoDefaultPlaybackSpeed.label, icon: Icons.speed, payload: videoDefaultPlaybackSpeed),
+                    options: [
+                      ListPickerItem(label: VideoPlayBackSpeed.pointTow5x.label, payload: VideoPlayBackSpeed.pointTow5x),
```
Comment (micahmo) [inline lib/settings/pages/video_player_settings.dart:170]: Can you try to add icons here too? Even if they're all the same? I think it looks better in the bottom modal.

File: pubspec.yaml [hunk 8]
```diff
@@ -113,16 +115,19 @@ dependencies:
...
 dev_dependencies:
   build_runner: ^2.4.6
   flutter_test:
     sdk: flutter
   flutter_lints: ^3.0.1
-  flutter_launcher_icons:
-    git:
-      url: https://github.com/fluttercommunity/flutter_launcher_icons.git
-      ref: master
+  # flutter_launcher_icons:
```
Comment (micahmo) [inline pubspec.yaml:127]: Why is this commented out?

**Review by micahmo — COMMENTED** [review]
Hey, I just wanted to say, this is super awesome, nice work! I especially love all the new settings! I think this is going to be a big hit.

I left some comments on the code, and I also had some overall thoughts after testing.
* It seems like the preview of videos with no image from Lemmy are completely blank. I think we should at least have some icon to stand in, like the web icon for URLs with no preview. However, now that we have an in-app video player, we can use a video icon instead of the globe icon.
* Some videos like this one never play. If there's an easy way to figure out which videos we support and which ones we don't, we can show the latter ones in the browser.
	* https://citizensgaming.com/post/751099
* From your demo, it looks like some videos play with a white background. Is there any way to make that black?

File: pubspec.yaml (hunk 8 above)
Comment (ggichure) [reply pubspec.yaml:127]: I keep having problems with `pub get`.

**Comment by ggichure** [comment]
> It seems like the preview of videos with no image from Lemmy are completely blank. I think we should at least have some icon to stand in, like the web icon for URLs with no preview. However, now that we have an in-app video player, we can use a video icon instead of the globe icon.

Resolved.

> From your demo, it looks like some videos play with a white background. Is there any way to make that black?

Resolved

> Some videos like this one never play. If there's an easy way to figure out which videos we support and which ones we don't, we can show the latter ones in the browser. 

At the moment , I have no way of checking for this.

**Comment by hjiangsu** [comment]
Once again, thanks for working on this! I'll be testing it out over the next few days to see if I can find any issues running it. I'll do a code review of it once I've tested it out 😄

**Comment by hjiangsu** [comment]
Just did some quick testing on this and there are a couple of things I'd like to adjust, but they might be better suited for separate PRs given that this one is already quite large!

What are your thoughts on merging this in first, and then applying additional PRs in the future? @micahmo

**Comment by micahmo** [comment]
I guess it depends on what you had in mind (whether the user experience will be poor without your suggestions, or whether they're more like enhancements). I think it also depends on whether you're planning a full release soon. But I'm totally fine with getting this in now. 😊

**Comment by hjiangsu** [comment]
The main thing I had in mind was to adjust the logic to work alongside `parsePostView` and `MediaView` so that all the post parsing logic is centralized! I did a quick code review and I think the core implementation of the video player (video player widgets, video settings, etc.) are good, so there shouldn't be many changes coming from there!

By centralizing the logic to `parsePostView`, this might also help reduce the instances where we have this issue:

> Some videos like this one never play. If there's an easy way to figure out which videos we support and which ones we don't, we can show the latter ones in the browser.

Of course, @ggichure let me know if you'd prefer to work on those changes (in which case, I can go into some more details about them) 😅

> I think it also depends on whether you're planning a full release soon.

I'd like to get initial/experimental video support in for the next full release, but I don't have a timeline for when the full release will be. I don't think there is a rush to get a full release soon either since the current version should already be compatible with Lemmy 0.19.4!

Thoughts? @micahmo @ggichure

**Comment by micahmo** [comment]
Agreed on all points! 😊

**Comment by ggichure** [comment]
@hjiangsu ,Sure,  do go ahead with the proposed changes.

**Comment by micahmo** [comment]
BTW, since this PR introduces new settings, keep in mind it'll have conflicts with #1348. I don't really care which ones goes first. 😊

**Review by hjiangsu — APPROVED** [review]

**Comment by hjiangsu** [comment]
I'll go ahead and merge this in first!

**Pull request merged**